---
title: "An Introduction to RbowtieCuda"
author: "Franck RICHARD"
date: '`r Sys.Date()`'
output:
  html_document: default
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{An Introduction to RbowtieCuda}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

NVBIO is a library of reusable components designed by NVIDIA Corporation to accelerate bioinformatics applications using CUDA.

Though it is specifically designed to unleash the power of NVIDIA GPUs, most of its components are completely cross-platform and can be used both from host C++ and device CUDA code.

The purpose of NVBIO is twofold: it can be thought of both as a solid basis to build new, modern applications targeting GPUs, which deferring the core computations to a library will always automatically and transparently benefit from new advances in GPU computing, and as example material to design novel bioinformatics algorithms for massively parallel architectures.

Additionally, NVBIO contains a suite of applications built on top of it, including a re-engineered implementation of the famous Bowtie2[1] short read aligner. Unlike many prototypes, nvBowtie is an attempt to build an industrial strength aligner, reproducing most of Bowtie2's original features as well as adding a few more, such as efficient support for direct BAM (and soon CRAM) output.

## Performances

nvBowtie is designed to exploit all the massive parallelism of modern GPUs, thus enabling a much higher alignment throughput at equal accuracy, or higher accuracy in the same time. Here's a graph showing nvBowtie's performance compared to bowtie2 on an Illumina HiSeq 2000 dataset (the first 10M reads of ERR161544) and an IonProton run, using both end-to-end and local alignment. The alignment results show 99.98% agreement at high MAPQ. All bowtie2 tests were run using 20 CPU threads, and default aligment options:

```{r}
knitr::include_graphics("benchmark-nvbowtie-speedup.png")
```


### BowtieCuda

This package provides an R wrapper of [nvBio/nvBowtie](https://github.com/NVlabs/nvbio). We wrap it into an R package that provide user friendly interfaces for R users. For more efficiency, the indexing and alignment functions (nvBWT and nvBowtieCuda) are written in C++ and are called thanks to the system2 function provided by the R language. All this in a transparent way for the user. The features offered by this package are therefore available in a version that is highly optimized for your machine.

An NVidia graphics card is required.


## Additional Installation Instructions

Please consult the INSTALL file.


## An Example Workflow by Using RbowtieCuda

### Installation

To install the latest version of RbowtieCuda, you will need to be using the latest version of R. RbowtieCuda is part of Bioconductor project, so you can install RbowtieCuda and its dependencies like this:

```
library(BiocManager)
BiocManager::install("RbowtieCuda")
```

Just like other R package, you need to load RbowtieCuda like this each time before using the package.

```{r loading}
library(RbowtieCuda)
```

### nvBWT : builds the BWT indices of the reference FASTA files

nvBWT is an application built on top of NVBIO to perform BWT-based reference indexing for nvBowtie and potentially other FM-index based applications. Given one or multiple fasta files, it will create both the forward and reverse BWT as well as 2-bit packed representation of the files and a few other auxiliary indices:

#### Example:

```{r idad}
td <- tempdir()
fa_file <- system.file(package="RbowtieCuda", "extdata", "bt2", "refs", "lambda_virus.fa")
nvBWT(myinput=fa_file, output=file.path(td, "index"), options="")
```

will generate the following files:

```
index.pac
index.rpac
index.bwt
index.rbwt
index.sa
index.rsa
index.ann
index.amb
```

nvBWT supports the following command options:
```
-v       | --verbosity     int (0-6) [5]     // select the verbosity level
-m       | --max-length    int       [inf]   // clamp input length
-b       | --byte-packing                    // output a byte-encoded .pac file
-w       | --word-packing                    // output a word-encoded .wpac file (more efficient)
-c       | --crc                             // compute CRCs
-d       | --device                          // select a cuda device
```

### nvBowtie Alignment:

nvBowtie is a GPU-accelerated re-engineering of Bowtie2, a very widely used short-read aligner. While being completely rewritten from scratch, nvBowtie reproduces many (though not all) of the features of Bowtie2.

nvBowtie is designed to exploit all the massive parallelism of modern GPUs, thus enabling a much higher alignment throughput at equal accuracy, or higher accuracy in the same time. While targeting maximum performance, nvBowtie is designed to match Bowtie2 as closely as possible, with the explicit goal of mantaining the same specificity and sensitivity characteristics. 

In order to take advantage of the massive parallelism available in modern processor architectures, nvBowtie re-implements the same underlying algorithms as Bowtie2 taking a fundamentally different approach. In fact, while Bowtie2 is essentially designed to operate on a single read at a time (possibly having multiple CPU threads working on a different read), carrying the entire alignment process in what is basically a very complex chain of nested function calls, nvBowtie works at all times with large batches of reads, and treats their alignment as a complex pipeline composed by many relatively simple but deeply parallel stages. In many of these stages parallelism is spread at a much finer granularity than at the reads-level, for example by processing many candidate hits from each read at the same time.


#### Example :

Reads_1 and reads_2 are raw paired-end reads file with fastq format. With bowtie2 index, reads will be mapped to reference by calling bowtie2. The result is saved in a bam file whose path is set to output.


```{r bt2bd1}
read_1 <- system.file(package="RbowtieCuda", "extdata", "bt2", "reads", "reads_1.fastq")
read_2 <- system.file(package="RbowtieCuda", "extdata", "bt2", "reads", "reads_2.fastq")
nvBowtie(file.path(td, "index"), file.path(td, "my_result.bam"), options="", seq1=read_1, seq2=read_2)
```

#### Indexing of bam files produced by nvBowtie

NvBowtie does not generate .bai index files associated with .bam files

These index files are essential in order to display a .bam file in the Integrative Genomics Viewer (IGV) program, for example.

However, this can be easily fixed. We will use the Rsamtools package which contains the necessary functions. If we have produced a results.bam file, for example...

It will be just enough to do :

```
library(Rsamtools) 
sortBam("results.bam", "results")
indexBam("results.bam")
```

#### Options and Version of Bowtie2 Aligner

It is possible to influence alignment using the following options :

```{r bt2usage}
nvBowtie_usage()
```

You can get version information by call:

```{r bt2version}
nvBowtie_version()
```

## Session Information

```{r sessioninfo}
sessionInfo()
```


## Acknowledgement

We would like to thank Ismael Galve Roperh for his assistance.


## Credits

The main contributors of the original NVBIO are:

    Jacopo Pantaleoni - jpantaleoni@nvidia.com
    Nuno Subtil - nsubtil@nvidia.com

RbowtieCuda developers:

    Samuel Simon Sanchez - samsimon@ucm.es
    Franck RICHARD - franck.richard@winstars.net

The maintainer of the RbowtieCuda package is [Franck RICHARD](mailto:franck.richard@winstars.net)


## Website

A small website containing help for the RbowtieCuda package can be found [here](https://belacqua-labo.ovh/bioinformatic/RbowtieCuda), and a github copy [here](https://github.com/FranckRICHARD01/RbowtieCuda)


## References

[1] Langmead, B., & Salzberg, S. L. (2012). Fast gapped-read alignment with Bowtie 2. Nature methods, 9(4), 357-359.

