<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>An Introduction to RbowtieCuda • RbowtieCuda</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="An Introduction to RbowtieCuda">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">RbowtieCuda</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/RbowtieCuda-Introduction.html">An Introduction to RbowtieCuda</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/FranckRICHARD01/RbowtieCuda/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>An Introduction to RbowtieCuda</h1>
                        <h4 data-toc-skip class="author">Franck
RICHARD</h4>
            
            <h4 data-toc-skip class="date">2024-12-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FranckRICHARD01/RbowtieCuda/blob/HEAD/vignettes/RbowtieCuda-Introduction.Rmd"><code>vignettes/RbowtieCuda-Introduction.Rmd</code></a></small>
      <div class="d-none name"><code>RbowtieCuda-Introduction.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>NVBIO is a library of reusable components developed by NVIDIA
Corporation to accelerate bioinformatics applications using CUDA
technology. Although specifically designed to harness the computational
power of NVIDIA GPUs, many of its components are fully cross-platform
and can be used in both host C++ and device CUDA code.</p>
<p>The purpose of NVBIO is twofold. It serves as a robust foundation for
developing modern GPU-focused applications, ensuring that core
computations delegated to the library automatically benefit from
advances in GPU technology. At the same time, it provides valuable
resources for designing novel bioinformatics algorithms tailored to
massively parallel architectures.</p>
<p>In addition to its core components, NVBIO includes a suite of
applications built on top of the library. Among these is nvBowtie, a
re-engineered implementation of the widely recognized Bowtie2 short-read
aligner <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Langmead, B., &amp;amp; Salzberg, S. L. (2012). Fast
gapped-read alignment with Bowtie 2. Nature methods, 9(4), 357-359.&lt;/p&gt;"><sup>1</sup></a>. Unlike many prototypes, nvBowtie is built
as an industrial-grade aligner, replicating most of Bowtie2’s original
features while adding enhancements such as efficient support for direct
BAM output, with future plans for CRAM support.</p>
</div>
<div class="section level2">
<h2 id="performances">Performances<a class="anchor" aria-label="anchor" href="#performances"></a>
</h2>
<p>nvBowtie is designed to fully exploit the massive parallelism of
modern GPUs, delivering significantly higher alignment throughput
without compromising accuracy, or achieving even greater accuracy in the
same amount of time. For example, its performance was compared to
Bowtie2 using an Illumina HiSeq 2000 dataset (the first 10 million reads
of ERR161544) and an IonProton dataset, applying both end-to-end and
local alignment methods. The results demonstrate an impressive 99.98%
agreement at high MAPQ scores. For comparison, Bowtie2 tests were
performed using 20 CPU threads with default alignment settings.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"benchmark-nvbowtie-speedup.png"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="benchmark-nvbowtie-speedup.png" alt="benchmark." width="610"><p class="caption">
benchmark
</p>
</div>
<div class="section level3">
<h3 id="rbowtiecuda">RBowtieCuda<a class="anchor" aria-label="anchor" href="#rbowtiecuda"></a>
</h3>
<p>This package provides an R wrapper for <a href="https://github.com/NVlabs/nvbio" class="external-link">nvBio/nvBowtie</a>, offering
user-friendly interfaces specifically designed for R users. To maximize
efficiency, the indexing and alignment functions (nvBWT and nvBowtie)
are implemented in C++ and seamlessly integrated into R using the
system2 function. This integration is fully transparent to the user,
ensuring that the package is easy to use while providing
high-performance features optimized to take full advantage of your
machine’s capabilities.</p>
<p>Additionally, an experimental implementation of the
<code>Wavefront Alignment (WFA)</code> method <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Marco-Sola S, Moure JC, Moreto M et al. Fast gap-affine
pairwise alignment using the wavefront algorithm. Bioinformatics
2021;37: 456–63.&lt;/p&gt;"><sup>2</sup></a> is included and can be
accessed by using the <code>--wfa</code> parameter during execution.</p>
<p>Please note that this package requires an NVIDIA graphics card for
proper functionality.</p>
</div>
</div>
<div class="section level2">
<h2 id="additional-installation-instructions">Additional Installation Instructions<a class="anchor" aria-label="anchor" href="#additional-installation-instructions"></a>
</h2>
<p>For detailed installation steps, please refer to the INSTALL file
included with the package.</p>
<p>RbowtieCuda is compatible with all Cuda versions greater than or
equal to 10.</p>
<p>However, compiling the nvBWT indexing tool results in an error when
using CUDA versions 12.5 or later. We are actively working on a
solution.</p>
</div>
<div class="section level2">
<h2 id="an-example-workflow-using-rbowtiecuda">An Example Workflow Using RbowtieCuda<a class="anchor" aria-label="anchor" href="#an-example-workflow-using-rbowtiecuda"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>To install the latest version of RbowtieCuda, ensure that you are
running the most up-to-date version of R. As RbowtieCuda is part of the
Bioconductor project, you can easily install it along with its
dependencies by following these steps:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"RbowtieCuda"</span><span class="op">)</span></span></code></pre>
<p>As with any other R package, you need to load RbowtieCuda each time
before using it, like this:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FranckRICHARD01/RbowtieCuda">RbowtieCuda</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biobase'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="nvbwt-building-bwt-indices-for-reference-fasta-files">nvBWT : Building BWT Indices for Reference FASTA Files<a class="anchor" aria-label="anchor" href="#nvbwt-building-bwt-indices-for-reference-fasta-files"></a>
</h3>
<p>nvBWT is an application developed as part of the NVBIO library,
designed to perform BWT-based reference indexing for nvBowtie and
potentially other FM-index-based applications. When provided with one or
more FASTA files, nvBWT generates both the forward and reverse
Burrows-Wheeler Transform (BWT), along with a 2-bit packed
representation of the sequences. In addition, it produces several
auxiliary indices to support efficient alignment and querying.</p>
<div class="section level4">
<h4 id="example">Example:<a class="anchor" aria-label="anchor" href="#example"></a>
</h4>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fa_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"RbowtieCuda"</span>, <span class="st">"extdata"</span>, <span class="st">"bt2"</span>, <span class="st">"refs"</span>, <span class="st">"lambda_virus.fa"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nvBWT.html">nvBWT</a></span><span class="op">(</span>myinput<span class="op">=</span><span class="va">fa_file</span>, output<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">"index"</span><span class="op">)</span>, options<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0</span></span></code></pre>
<p>will generate the following files:</p>
<pre><code><span><span class="va">index.pac</span></span>
<span><span class="va">index.rpac</span></span>
<span><span class="va">index.bwt</span></span>
<span><span class="va">index.rbwt</span></span>
<span><span class="va">index.sa</span></span>
<span><span class="va">index.rsa</span></span>
<span><span class="va">index.ann</span></span>
<span><span class="va">index.amb</span></span></code></pre>
<p>nvBWT supports the following command options:</p>
<pre><code>-v       | --verbosity     int (0-6) [5]     // select the verbosity level
-m       | --max-length    int       [inf]   // clamp input length
-b       | --byte-packing                    // output a byte-encoded .pac file
-w       | --word-packing                    // output a word-encoded .wpac file (more efficient)
-c       | --crc                             // compute CRCs
-d       | --device                          // select a cuda device</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="nvbowtie-alignments">nvBowtie Alignments:<a class="anchor" aria-label="anchor" href="#nvbowtie-alignments"></a>
</h3>
<p>nvBowtie is a GPU-accelerated re-engineering of Bowtie2, one of the
most widely used short-read aligners. Completely rewritten from scratch,
nvBowtie retains many of the key features of Bowtie2, though not all
functionalities are replicated.</p>
<p>Designed to fully exploit the massive parallelism of modern GPUs,
nvBowtie achieves significantly higher alignment throughput without
sacrificing accuracy—or offers even greater accuracy within the same
time frame. Despite its focus on performance, nvBowtie is carefully
designed to align closely with Bowtie2 in terms of specificity and
sensitivity, maintaining the same level of reliability for users.</p>
<p>To harness the computational power of modern processor architectures,
nvBowtie re-implements the algorithms underlying Bowtie2 but adopts a
fundamentally different approach. While Bowtie2 is optimized to process
one read at a time—using multiple CPU threads to handle different reads
simultaneously—nvBowtie operates on large batches of reads, treating
their alignment as a pipeline. This pipeline consists of many relatively
simple but highly parallel stages, each optimized for execution on GPUs.
In several stages, the parallelism extends far beyond the read level,
processing multiple candidate hits for each read simultaneously,
enabling a much finer granularity of parallel computation.</p>
<div class="section level4">
<h4 id="new-features">New features :<a class="anchor" aria-label="anchor" href="#new-features"></a>
</h4>
<p>We’ve introduced several new functions to nvBowtie. You can now
perform alignments using the WFA method by including the
<code>--wfa  (or --scoring wfa)</code> parameter. The WFA method
requires a large amount of RAM on the graphics card. We therefore
recommend using an Nvidia card with 8GB or more. Please note that this
feature is still experimental; it currently supports only end-to-end
alignments and does not yet allow customization of scoring parameters.
By default, it uses the following scoring: match:0, mismatch:1,
gap_open:1 and gap_ext:1.</p>
<p>Additionally, the <code>--cache-writes</code> parameter optimizes
disk write operations, resulting in faster alignments. This
functionality requires 4GB of RAM and is limited to paired-end
alignments.</p>
</div>
<div class="section level4">
<h4 id="example-1">Example :<a class="anchor" aria-label="anchor" href="#example-1"></a>
</h4>
<p>Reads_1 and Reads_2 represent raw paired-end read files in FASTQ
format. Using a nvBWT index, these reads are mapped to the reference
genome by invoking nvBowtie. The resulting alignments are stored in a
BAM file, with its file path specified by the output parameter.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">read_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"RbowtieCuda"</span>, <span class="st">"extdata"</span>, <span class="st">"bt2"</span>, <span class="st">"reads"</span>, <span class="st">"reads_1.fastq"</span><span class="op">)</span></span>
<span><span class="va">read_2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package<span class="op">=</span><span class="st">"RbowtieCuda"</span>, <span class="st">"extdata"</span>, <span class="st">"bt2"</span>, <span class="st">"reads"</span>, <span class="st">"reads_2.fastq"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/nvBowtie.html">nvBowtie</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">"index"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">"my_result.bam"</span><span class="op">)</span>, options<span class="op">=</span><span class="st">""</span>, seq1<span class="op">=</span><span class="va">read_1</span>, seq2<span class="op">=</span><span class="va">read_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="indexing-of-bam-files-produced-by-nvbowtie">Indexing of bam files produced by nvBowtie<a class="anchor" aria-label="anchor" href="#indexing-of-bam-files-produced-by-nvbowtie"></a>
</h4>
<p>nvBowtie does not automatically generate the .bai index files that
are typically associated with .bam files.</p>
<p>These index files are essential for visualizing .bam files in tools
such as the Integrative Genomics Viewer (IGV).</p>
<p>Fortunately, this issue can be easily resolved using the Rsamtools
package, which includes the required functionality. For example, if you
have generated a file named results.bam, you can create the
corresponding index file with a simple command.</p>
<p>You only need to run the following in R:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Rsamtools" class="external-link">Rsamtools</a></span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">sortBam</a></span><span class="op">(</span><span class="st">"results.bam"</span>, <span class="st">"results"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/scanBam.html" class="external-link">indexBam</a></span><span class="op">(</span><span class="st">"results.bam"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="section"></h4>
</div>
<div class="section level4">
<h4 id="options-and-version-of-the-nvbowtie-aligner">Options and Version of the nvBowtie Aligner<a class="anchor" aria-label="anchor" href="#options-and-version-of-the-nvbowtie-aligner"></a>
</h4>
<p>You can customize the alignment process by adjusting the available
options, enabling you to optimize performance and accuracy according to
your specific needs:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nvBowtie_usage.html">nvBowtie_usage</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "options:"</span></span>
<span><span class="co">## [1] "General:"</span></span>
<span><span class="co">## [1] "  --verbosity         int                    [5]        verbosity level"</span></span>
<span><span class="co">## [1] "  --upto       | -u   int                    [-1]       maximum number of reads to process"</span></span>
<span><span class="co">## [1] "  --trim3      | -3   int                    [0]        trim the first N bases of 3'"</span></span>
<span><span class="co">## [1] "  --trim5      | -5   int                    [0]        trim the first N bases of 5'"</span></span>
<span><span class="co">## [1] "  --nofw                                     [false]    do not align the forward strand"</span></span>
<span><span class="co">## [1] "  --norc                                     [false]    do not align the reverse-complemented strand"</span></span>
<span><span class="co">## [1] "  --device            int                    [0]        select the given cuda device(s) (e.g. --device 0 --device 1 ...)"</span></span>
<span><span class="co">## [1] "  --file-ref                                 [false]    load reference from file"</span></span>
<span><span class="co">## [1] "  --server-ref                               [false]    load reference from server"</span></span>
<span><span class="co">## [1] "  --phred33                                  [true]     qualities are ASCII characters equal to Phred quality + 33"</span></span>
<span><span class="co">## [1] "  --phred64                                  [false]    qualities are ASCII characters equal to Phred quality + 64"</span></span>
<span><span class="co">## [1] "  --solexa-quals                             [false]    qualities are in the Solexa format"</span></span>
<span><span class="co">## [1] "  --rg-id             string                            add the RG-ID field of the SAM output header"</span></span>
<span><span class="co">## [1] "  --rg                string,val                        add an RG-TAG field of the SAM output header"</span></span>
<span><span class="co">## [1] "  --cache-writes      bool                   [false]    speed up writes on disk"</span></span>
<span><span class="co">## [1] "Paired-End:"</span></span>
<span><span class="co">## [1] "  --ff                                       [false]    paired mates are forward-forward"</span></span>
<span><span class="co">## [1] "  --fr                                       [true]     paired mates are forward-reverse"</span></span>
<span><span class="co">## [1] "  --rf                                       [false]    paired mates are reverse-forwardd"</span></span>
<span><span class="co">## [1] "  --rr                                       [false]    paired mates are reverse-reverse"</span></span>
<span><span class="co">## [1] "  --minins     |  -I  int                    [0]        minimum insert length"</span></span>
<span><span class="co">## [1] "  --maxins     |  -X  int                    [500]      maximum insert length"</span></span>
<span><span class="co">## [1] "  --overlap                                  [true]     allow overlapping mates"</span></span>
<span><span class="co">## [1] "  --no-mixed                                 [false]    only report paired alignments"</span></span>
<span><span class="co">## [1] "  --ungapped-mates | -ug                                perform ungapped mate alignment"</span></span>
<span><span class="co">## [1] "Seeding:"</span></span>
<span><span class="co">## [1] "  --seed-len   | -L   int                    [22]       seed lengths"</span></span>
<span><span class="co">## [1] "  --seed-freq  | -i   {G|L|S},x,y                       seed interval, as x + y*func(read-len) (G=log,L=linear,S=sqrt)"</span></span>
<span><span class="co">## [1] "  --max-hits          int                    [100]      maximum amount of seed hits"</span></span>
<span><span class="co">## [1] "  --max-reseed | -R   int                    [2]        number of reseeding rounds"</span></span>
<span><span class="co">## [1] "  --top               bool                   [false]    explore top seed entirely"</span></span>
<span><span class="co">## [1] "  --N                 bool                   [false]    allow substitution in seed"</span></span>
<span><span class="co">## [1] "Extension:"</span></span>
<span><span class="co">## [1] "  --mode              {best,best-exact,all}  [best]     alignment mode\n"</span></span>
<span><span class="co">## [1] "  --all        | -a                          [false]    perform all-mapping (i.e. find and report all alignments)"</span></span>
<span><span class="co">## [1] "  --local                                    [false]    perform local alignment"</span></span>
<span><span class="co">## [1] "  --rand                                     [true]     randomized seed hit selection"</span></span>
<span><span class="co">## [1] "  --no-rand                                  [false]    do not randomize seed hit selection"</span></span>
<span><span class="co">## [1] "  --max-dist          int                    [15]       maximum edit distance"</span></span>
<span><span class="co">## [1] "  --max-effort-init   int                    [15]       initial maximum number of consecutive extension failures"</span></span>
<span><span class="co">## [1] "  --max-effort | -D   int                    [15]       maximum number of consecutive extension failures"</span></span>
<span><span class="co">## [1] "  --min-ext           int                    [30]       minimum number of extensions per read"</span></span>
<span><span class="co">## [1] "  --max-ext           int                    [400]      maximum number of extensions per read"</span></span>
<span><span class="co">## [1] "  --very-fast                                           apply the very-fast presets"</span></span>
<span><span class="co">## [1] "  --fast                                                apply the fast presets"</span></span>
<span><span class="co">## [1] "  --sensitive                                           apply the sensitive presets"</span></span>
<span><span class="co">## [1] "  --very-sensitive                                      apply the very-sensitive presets"</span></span>
<span><span class="co">## [1] "  --very-fast-local                                     apply the very-fast presets"</span></span>
<span><span class="co">## [1] "  --fast-local                                          apply the fast presets"</span></span>
<span><span class="co">## [1] "  --sensitive-local                                     apply the sensitive presets"</span></span>
<span><span class="co">## [1] "  --very-sensitive-local                                apply the very-sensitive presets"</span></span>
<span><span class="co">## [1] "Scoring:"</span></span>
<span><span class="co">## [1] "  --scoring           {sw|ed|wfa}            [ed]       Smith-Waterman / Edit-Distance / Wfa scoring"</span></span>
<span><span class="co">## [1] "  --score-min         {G|L|S},x,y                       minimum score function, as x + y*func(read-len)"</span></span>
<span><span class="co">## [1] "  --ma                int                               match bonus"</span></span>
<span><span class="co">## [1] "  --mp                int,int                           mismatch min/max penalties"</span></span>
<span><span class="co">## [1] "  --np                int                               N penalty"</span></span>
<span><span class="co">## [1] "  --rdg               int,int                           read open/extension gap penalties"</span></span>
<span><span class="co">## [1] "  --rfg               int,int                           reference open/extension gap penalties"</span></span>
<span><span class="co">## [1] "Alternative:"</span></span>
<span><span class="co">## [1] "  --wfa                                                 Activate wavefront algorithm"</span></span>
<span><span class="co">## [1] "Reporting:"</span></span>
<span><span class="co">## [1] "  --mapQ-filter | -Q  int                    [0]        minimum mapQ threshold"</span></span>
<span><span class="co">## [1] ""</span></span>
<span><span class="co">## [1] ""</span></span>
<span><span class="co">## [1] "Default values are indicated in brackets []."</span></span>
<span><span class="co">## [1] ""</span></span>
<span><span class="co">## [1] "The '--scoring-scheme filename' option allows to provide a custom Smith-Waterman scoring"</span></span>
<span><span class="co">## [1] "scheme through a text file, where each line must contain a token value pair."</span></span>
<span><span class="co">## [1] " The tokens and default values are reported below:"</span></span>
<span><span class="co">## [1] "*  match               0        // local alignment: 2"</span></span>
<span><span class="co">## [1] "*  mm-penalty-min      2"</span></span>
<span><span class="co">## [1] "*  mm-penalty-max      6"</span></span>
<span><span class="co">## [1] "*  N-penalty-min       1"</span></span>
<span><span class="co">## [1] "*  N-penalty-max       1"</span></span>
<span><span class="co">## [1] "*  score-min-const     -0.6     // local alignment: 0"</span></span>
<span><span class="co">## [1] "*  score-min-coeff     -0.6     // local alignment: 10"</span></span>
<span><span class="co">## [1] "*  score-min-type      linear   // local alignment: log"</span></span>
<span><span class="co">## [1] "*  N-ceil-const        0"</span></span>
<span><span class="co">## [1] "*  N-ceil-coeff        0.15"</span></span>
<span><span class="co">## [1] "*  read-gap-const      5"</span></span>
<span><span class="co">## [1] "*  read-gap-coeff      3"</span></span>
<span><span class="co">## [1] "*  ref-gap-const       5"</span></span>
<span><span class="co">## [1] "*  ref-gap-coeff       3"</span></span>
<span><span class="co">## [1] "*  gap-free            5"</span></span></code></pre>
<p>You can execute unit tests for the program using the following
commands or tools, ensuring its functionality and correctness:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nvbio_tests.html">nvbio_tests</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0</span></span></code></pre>
<p>And you can obtain version information by executing the following
command:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/nvBowtie_version.html">nvBowtie_version</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.10</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=fr_FR.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=es_ES.UTF-8        LC_COLLATE=fr_FR.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=es_ES.UTF-8    LC_MESSAGES=fr_FR.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=es_ES.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=es_ES.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Madrid</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] RbowtieCuda_0.1.2           SummarizedExperiment_1.36.0</span></span>
<span><span class="co">##  [3] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">##  [5] GenomeInfoDb_1.42.0         IRanges_2.40.0             </span></span>
<span><span class="co">##  [7] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">##  [9] MatrixGenerics_1.18.0       matrixStats_1.4.1          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] Matrix_1.7-0            jsonlite_1.8.9          compiler_4.4.1         </span></span>
<span><span class="co">##  [4] crayon_1.5.3            jquerylib_0.1.4         systemfonts_1.1.0      </span></span>
<span><span class="co">##  [7] textshaping_0.4.0       png_0.1-8               yaml_2.3.10            </span></span>
<span><span class="co">## [10] fastmap_1.2.0           lattice_0.22-6          R6_2.5.1               </span></span>
<span><span class="co">## [13] XVector_0.46.0          S4Arrays_1.6.0          knitr_1.49             </span></span>
<span><span class="co">## [16] htmlwidgets_1.6.4       DelayedArray_0.32.0     desc_1.4.3             </span></span>
<span><span class="co">## [19] GenomeInfoDbData_1.2.13 bslib_0.8.0             rlang_1.1.4            </span></span>
<span><span class="co">## [22] cachem_1.1.0            xfun_0.49               fs_1.6.5               </span></span>
<span><span class="co">## [25] sass_0.4.9              SparseArray_1.6.0       cli_3.6.3              </span></span>
<span><span class="co">## [28] pkgdown_2.1.1           zlibbioc_1.52.0         grid_4.4.1             </span></span>
<span><span class="co">## [31] digest_0.6.37           rstudioapi_0.17.1       lifecycle_1.0.4        </span></span>
<span><span class="co">## [34] evaluate_1.0.1          ragg_1.3.3              abind_1.4-8            </span></span>
<span><span class="co">## [37] rmarkdown_2.29          httr_1.4.7              tools_4.4.1            </span></span>
<span><span class="co">## [40] htmltools_0.5.8.1       UCSC.utils_1.2.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="acknowledgement">Acknowledgement<a class="anchor" aria-label="anchor" href="#acknowledgement"></a>
</h2>
<p>We would like to thank Ismael Galve Roperh for his assistance.</p>
</div>
<div class="section level2">
<h2 id="credits">Credits<a class="anchor" aria-label="anchor" href="#credits"></a>
</h2>
<p>The main contributors of the original NVBIO are:</p>
<pre><code>Jacopo Pantaleoni - jpantaleoni@nvidia.com
Nuno Subtil - nsubtil@nvidia.com</code></pre>
<p>RbowtieCuda developers:</p>
<pre><code>Samuel Simon Sanchez - samsimon@ucm.es
Franck RICHARD - franck.richard@winstars.net</code></pre>
<p>The maintainer of the RbowtieCuda package is <a href="mailto:franck.richard@winstars.net">Franck RICHARD</a></p>
</div>
<div class="section level2">
<h2 id="website">Website<a class="anchor" aria-label="anchor" href="#website"></a>
</h2>
<p>A dedicated website with helpful resources for the RbowtieCuda
package is available <a href="https://belacqua-labo.ovh/bioinformatic/RbowtieCuda" class="external-link">here</a>, and
a github copy <a href="https://github.com/FranckRICHARD01/RbowtieCuda">here</a></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>[1] Langmead, B., &amp; Salzberg, S. L. (2012). Fast gapped-read
alignment with Bowtie 2. Nature methods, 9(4), 357-359.</p>
<p>[2] Marco-Sola S, Moure JC, Moreto M et al. Fast gap-affine pairwise
alignment using the wavefront algorithm. Bioinformatics 2021;37:
456–63.</p>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jacopo Pantaleoni, Nuno Subtil, Samuel Simon-Sanchez, Franck RICHARD.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
